File: python_75/two/max_prod_subarr.py
# input = [2, 3, -2, 4, -4]
# output = 6

nums = [2, 3, -2, 4]

curProd = 1
maxProd = 1

for num in nums:
  if (curProd < 0 and num < 0) or (curProd > 0 and num > 0):
    curProd = curProd * num
  else: 
    curProd = 1
  maxProd = max(curProd, maxProd)

print(maxProd)

File: spring-data-redis-example/src/main/java/com/sunil/redis/ds/maxProdSubArr.java
package com.sunil.redis.ds;

public class maxProdSubArr {

    public static void main(String[] args) {
        int[] nums = {2, 3, -4, 3, -2, 2};
        System.out.println(maxProd(nums));
    }

    public static int maxProd(int[] nums) {

        int curMin = 1;
        int curMax = 1;
        int max = 1;

        for (int num : nums) {
            int temp = Math.max(curMax * num, curMin * num);
            curMin = Math.min(curMax * num, curMin * num);
            curMax = temp;
            max = Math.max(curMax, max);
        }
        return max;
    }
}
